### PR TITLE
fix(makefile): fix go-gpt2 folder and add verification before git clone

### PR DESCRIPTION
**Pull request by mkellerman**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -44,43 +44,43 @@ go-gpt4all-j/libgptj.a-generic: go-gpt4all-j
...
+	$(MAKE) -C go-gpt2 libgpt2.a
 
-go-gpt2.cpp/libgpt2.a-generic: go-gpt2.cpp
-	$(MAKE) -C go-gpt2.cpp generic-libgpt2.a
+go-gpt2/libgpt2.a-generic: go-gpt2
+	$(MAKE) -C go-gpt2 generic-libgpt2.a
 
 go-llama:
-	git clone -b $(GOLLAMA_VERSION) --recurse-submodules https://github.com/go-skynet/go-llama.cpp go-llama
+	if [ ! -d "$(shell pwd)/go-gpt4all-j" ]; then git clone -b $(GOLLAMA_VERSION) --recurse-submodules https://github.com/go-skynet/go-llama.cpp go-llama; fi
```
Comment (mudler) [inline Makefile:64]:
```suggestion
	if [ ! -d "$(shell pwd)/go-llama" ]; then git clone -b $(GOLLAMA_VERSION) --recurse-submodules https://github.com/go-skynet/go-llama.cpp go-llama; fi
```

File: Makefile [hunk 2]
```diff
@@ -44,43 +44,43 @@ go-gpt4all-j/libgptj.a-generic: go-gpt4all-j
...
 	$(MAKE) -C go-llama libbinding.a
 
 go-llama-generic:
-	git clone -b $(GOLLAMA_VERSION) --recurse-submodules https://github.com/go-skynet/go-llama.cpp go-llama
+	if [ ! -d "$(shell pwd)/go-gpt4all-j" ]; then git clone -b $(GOLLAMA_VERSION) --recurse-submodules https://github.com/go-skynet/go-llama.cpp go-llama; fi
```
Comment (mudler) [inline Makefile:68]:
```suggestion
	if [ ! -d "$(shell pwd)/go-llama" ]; then git clone -b $(GOLLAMA_VERSION) --recurse-submodules https://github.com/go-skynet/go-llama.cpp go-llama; fi
```

File: Makefile [hunk 3]
```diff
@@ -44,43 +44,43 @@ go-gpt4all-j/libgptj.a-generic: go-gpt4all-j
 	$(MAKE) -C go-gpt4all-j generic-libgptj.a
 
 # CEREBRAS GPT
-go-gpt2.cpp:
```
Comment (mudler) [inline Makefile:47]: I wonder, this PR is renaming go-gpt2.cpp to go-gpt2, is that for consistency, or somehow Mac didn't liked it?

File: Makefile [hunk 4]
```diff
@@ -19,14 +19,14 @@ all: help
...
+	C_INCLUDE_PATH=$(shell pwd)/go-llama:$(shell pwd)/go-gpt4all-j:$(shell pwd)/go-gpt2 LIBRARY_PATH=$(shell pwd)/go-llama:$(shell pwd)/go-gpt4all-j:$(shell pwd)/go-gpt2 $(GOCMD) build -o $(BINARY_NAME) ./
 
 buildgeneric: prepare-generic ## Build the project
-	C_INCLUDE_PATH=$(shell pwd)/go-llama.cpp:$(shell pwd)/go-gpt4all-j:$(shell pwd)/go-gpt2.cpp LIBRARY_PATH=$(shell pwd)/go-llama.cpp:$(shell pwd)/go-gpt4all-j:$(shell pwd)/go-gpt2.cpp $(GOCMD) build -o $(BINARY_NAME) ./
+	C_INCLUDE_PATH=$(shell pwd)/go-llama:$(shell pwd)/go-gpt4all-j:$(shell pwd)/go-gpt2 LIBRARY_PATH=$(shell pwd)/go-llama:$(shell pwd)/go-gpt4all-j:$(shell pwd)/go-gpt2 $(GOCMD) build -o $(BINARY_NAME) ./
 
 ## GPT4ALL-J
 go-gpt4all-j:
-	git clone --recurse-submodules https://github.com/go-skynet/go-gpt4all-j.cpp go-gpt4all-j && cd go-gpt4all-j && git checkout -b build $(GOGPT4ALLJ_VERSION)
+	if [ ! -d "$(shell pwd)/go-gpt4all-j" ]; then git clone --recurse-submodules https://github.com/go-skynet/go-gpt4all-j.cpp go-gpt4all-j && cd go-gpt4all-j && git checkout -b build $(GOGPT4ALLJ_VERSION); fi
```
Comment (mudler) [inline Makefile:29]: the if seems redundant here, when a make target is named after a file or a directory, it shouldn't re-execute again as the dir is already populated:

```
base ❯ make go-llama
.....

g++ -I. -I./examples -O3 -DNDEBUG -std=c++11 -fPIC -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wno-multichar -pthread -march=native -mtune=native -c llama.cpp -o llama.o
make[2]: Leaving directory '/home/mudler/_git/LocalAI/go-llama/llama.cpp'
make -C llama.cpp common.o
make[2]: Entering directory '/home/mudler/_git/LocalAI/go-llama/llama.cpp'
I llama.cpp build info: 
I UNAME_S:  Linux
I UNAME_P:  x86_64
I UNAME_M:  x86_64
I CFLAGS:   -I.              -O3 -DNDEBUG -std=c11   -fPIC -Wall -Wextra -Wpedantic -Wcast-qual -Wdouble-promotion -Wshadow -Wstrict-prototypes -Wpointer-arith -pthread -march=native -mtune=native
I CXXFLAGS: -I. -I./examples -O3 -DNDEBUG -std=c++11 -fPIC -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wno-multichar -pthread -march=native -mtune=native
I LDFLAGS:  
I CC:       cc (SUSE Linux) 12.2.1 20230124 [revision 193f7e62815b4089dfaed4c2bd34fd4f10209e27]
I CXX:      g++ (SUSE Linux) 12.2.1 20230124 [revision 193f7e62815b4089dfaed4c2bd34fd4f10209e27]

g++ -I. -I./examples -O3 -DNDEBUG -std=c++11 -fPIC -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wno-multichar -pthread -march=native -mtune=native -c examples/common.cpp -o common.o
make[2]: Leaving directory '/home/mudler/_git/LocalAI/go-llama/llama.cpp'
g++ -I./llama.cpp -I. -I./llama.cpp/examples -I./examples -O3 -DNDEBUG -std=c++11 -fPIC -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -pthread -I./llama.cpp -I./llama.cpp/examples binding.cpp -o binding.o -c 
ar src libbinding.a llama.cpp/ggml.o llama.cpp/common.o llama.cpp/llama.o binding.o
make[1]: Leaving directory '/home/mudler/_git/LocalAI/go-llama'

~/_git/LocalAI use_default_model* 15s
base ❯ make go-llama
make: 'go-llama' is up to date.

Comment (mudler) [reply Makefile:29]: what I'm missing here?

File: .gitignore [hunk 5]
```diff
@@ -1,6 +1,7 @@
 # go-llama build artifacts
 go-llama
 go-gpt4all-j
+go-gpt2
```
Comment (mudler) [inline .gitignore:4]: good catch!

File: Makefile (hunk 3 above)
Comment (mkellerman) [reply Makefile:47]: consistancy

File: Makefile (hunk 4 above)
Comment (mkellerman) [reply Makefile:29]: that behaviour doesn't see to happen for me. will try and get back to you asap.

Comment (mkellerman) [reply Makefile:29]: You're right.. because i was using generics, the paths dont align.. so implementing a fix now.

File: docker-compose.yaml [hunk 6]
```diff
@@ -2,18 +2,14 @@ version: '3.6'
 
 services:
   api:
-    image: quay.io/go-skynet/local-ai:latest
+    # image: quay.io/go-skynet/local-ai:latest
```
Comment (mudler) [inline docker-compose.yaml:5]:
```suggestion
     image: quay.io/go-skynet/local-ai:latest
```

File: .env [hunk 7]
```diff
@@ -1,4 +1,5 @@
 THREADS=14
 CONTEXT_SIZE=512
 MODELS_PATH=/models
-# DEBUG=true
\ No newline at end of file
+DEBUG=true
+BUILD_TYPE=generic
```
Comment (mudler) [inline .env:5]:
```suggestion
#BUILD_TYPE=generic
```

File: .env [hunk 8]
```diff
@@ -1,4 +1,5 @@
 THREADS=14
 CONTEXT_SIZE=512
 MODELS_PATH=/models
-# DEBUG=true
\ No newline at end of file
+DEBUG=true
```
Comment (mudler) [inline .env:4]:
```suggestion
#DEBUG=true
```

File: docker-compose.yaml [hunk 9]
```diff
@@ -2,18 +2,14 @@ version: '3.6'
 
 services:
   api:
-    image: quay.io/go-skynet/local-ai:latest
+     image: quay.io/go-skynet/local-ai:latest
```
Comment (mudler) [inline docker-compose.yaml:5]:
```suggestion
    image: quay.io/go-skynet/local-ai:latest
```

**Review by mudler — APPROVED** [review]

**Comment by mudler** [comment]
Awesome job @mkellerman !

**Pull request merged**